### PR TITLE
PAYEE-594: Facilitator to drop the cache for acceptance tests

### PR DESCRIPTION
**Pull request by vivekhmrc**
- [ ] Unit tests passing
- [ ] Coverage reached
- [ ] Acceptance tests passing
- [ ] Reviewed

### CODE REVIEW

File: app/uk/gov/hmrc/tai/connectors/JourneyCacheConnector.scala [hunk 1]
```diff
@@ -84,4 +84,8 @@ class JourneyCacheConnector @Inject()(httpHandler: HttpHandler, servicesConfig:
   def flushWithEmpId(journeyName: String, empId: Int)(implicit hc: HeaderCarrier): Future[TaiResponse] =
     httpHandler.deleteFromApi(cacheUrl(s"$journeyName/$empId")).map(_ => TaiSuccessResponse)
 
+  def testOnlyCacheUrl(journeyName: String) = s"$serviceUrl/tai/test-only/journey-cache/$journeyName"
```
Comment (brendanstraughan) [inline app/uk/gov/hmrc/tai/connectors/JourneyCacheConnector.scala:87]: why do we need to make a new route, why cant we just use the existing method for flushing the cache?

File: app/uk/gov/hmrc/tai/service/journeyCache/JourneyCacheService.scala [hunk 2]
```diff
@@ -133,4 +134,7 @@ class JourneyCacheService @Inject()(val journeyName: String, journeyCacheConnect
 
   def flushWithEmpId(empId: Int)(implicit hc: HeaderCarrier): Future[TaiResponse] =
```
Comment (brendanstraughan) [inline app/uk/gov/hmrc/tai/service/journeyCache/JourneyCacheService.scala:135]: can we not just use this?

File: conf/testOnlyDoNotUseInAppConf.routes [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+# To do so add the following line to the micro-service profile: "-Dapplication.router=testOnlyDoNotUseInAppConf.Routes"
+# To start the micro-service locally using the test routes run the following command: "sbt run -Dapplication.router=testOnlyDoNotUseInAppConf.Routes" 
+
+# Any test-only end-point should be defined here.
+# !!!WARNING!!! Every route defined in this file MUST be prefixed with "/test-only/". This is because NGINX is blocking every uri containing the string "test-only" in production.
+# Failing to follow this rule may result in test routes deployed in production.
+
+# Add all the application routes to the prod.routes file
+# microservice specific routes
+
```
Comment (brendanstraughan) [inline conf/testOnlyDoNotUseInAppConf.routes:14]: why do we need this file, are there any routes being used in here?

File: conf/tai.routes [hunk 4]
```diff
@@ -165,6 +165,7 @@ POST        /update-income/bonus-overtime-amount/:empId
 ## Regular Income URIs
 GET         /update-income/cancel/:id                                                  controllers.IncomeController.cancel(id: Int)
 GET         /update-income/select-taxable-pay                                          controllers.IncomeController.viewIncomeForEdit
+GET         /test-only/update-income/delete/:id                                        controllers.IncomeController.delete(id: Int)
```
Comment (sergi-hmrc) [inline conf/tai.routes:168]: why not just using the DELETE verb?

Comment (sergi-hmrc) [reply conf/tai.routes:168]: Also this should be part of conf/testOnlyDoNotUseInAppConf.routes

**Comment by sergi-hmrc** [comment]
A few things: 
Rather than creating `conf/testOnlyDoNotUseInAppConf.routes` we should reuse `dev.routes`. 
Anything inside `dev.routes` will need a different package notice the `controllers.testOnly`: 
`# Routes for development environments
DELETE         /update-income/:id   controllers.testOnly.IncomeController.delete(id: Int)
`
As described [here](https://stackoverflow.com/questions/20211734/play-framework-split-routes-in-multiple-files-without-sub-projects) it won't work otherwise.

Finally, if the only intention is to delete the cache for testing purposes I don't think we should redirect the user to any page but just do a DELETE against the endpoint and return an OK/No Content etc

File: app/uk/gov/hmrc/tai/service/journeyCache/JourneyCacheService.scala (hunk 2 above)
Comment (vivekhmrc) [reply app/uk/gov/hmrc/tai/service/journeyCache/JourneyCacheService.scala:124]: This is just to keep the test ones separate

File: conf/testOnlyDoNotUseInAppConf.routes (hunk 3 above)
Comment (vivekhmrc) [reply conf/testOnlyDoNotUseInAppConf.routes:14]: Yes moved the test only route here

File: conf/tai.routes (hunk 4 above)
Comment (vivekhmrc) [reply conf/tai.routes]: Get is to facilitate testers. The selenium web driver can easily call Get as its doing at other places. With Delete the testers will have to write custom code for this.

File: app/uk/gov/hmrc/tai/connectors/JourneyCacheConnector.scala (hunk 1 above)
Comment (vivekhmrc) [reply app/uk/gov/hmrc/tai/connectors/JourneyCacheConnector.scala:76]: This is just to keep the test ones separate

File: app/controllers/testOnly/TaiUpdateIncomeController.scala [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+class TaiUpdateIncomeController @Inject()(
+  @Named("Update Income") journeyCacheService: JourneyCacheService,
+  authenticate: AuthAction,
+  validatePerson: ValidatePerson,
+  mcc: MessagesControllerComponents,
+  implicit val templateRenderer: TemplateRenderer
+)(implicit ec: ExecutionContext)
+    extends TaiBaseController(mcc) {
+
+  def delete(empId: Int): Action[AnyContent] = (authenticate andThen validatePerson).async { implicit request =>
```
Comment (brendanstraughan) [inline app/controllers/testOnly/TaiUpdateIncomeController.scala:40]: I thought we said we were only going to make changes to the backend
@vivekhmrc @sergi-hmrc

Comment (sergi-hmrc) [reply app/controllers/testOnly/TaiUpdateIncomeController.scala:40]: Agree ^

Comment (vivekhmrc) [reply app/controllers/testOnly/TaiUpdateIncomeController.scala:40]: In the current tai-acceptance-tests we don't have a Web Service Client to directly invoke **tai**.  So we need a test route here to facilitate it. Once the **_tai-acceptance-test_** are enhanced to have a Web service client e.g. Akka HTTP or some other then we can remove this. It's captured here

https://jira.tools.tax.service.gov.uk/browse/PAYEE-872

**Pull request merged**